security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,187 @@
+	}
+	return c, nil
+}
+
+func responseTimeTrack(start time.Time, name string) {
+	elapsed := time.Since(start)
+	cLog.Debugf("%s took %s", name, elapsed)
+}
+
+// CreateCertificate is similar to Sign but returns the leaf cert and the entire cert chain.
+func (c *CAClient) CreateCertificate(ctx context.Context,
+	req *pb.IstioCertificateRequest, identities []string) (*pb.IstioCertificateResponse, error) {
+	if err := verifyCSRIdentities(identities, req.GetCsr()); err != nil {
+		return nil, fmt.Errorf("verify CSR Identities failed: %v", err)
+	}
+	cLog.Debugf("forwarding Workload's CSR to Custom CA server...")
+	resp, err := c.signingAPIClient.CreateCertificate(ctx, req)
+	certChain := resp.GetCertChain()
+	if err != nil {
+		cLog.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+		return nil, fmt.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+	}
+
+	var responseCertChains []string
+	certChainWithoutRoot := certChain[:len(certChain)-1]
+	for _, cert := range certChainWithoutRoot {
+		parsedCert, err := validateAndParseCert(cert)
+		if err != nil {
+			cLog.Errorf("response certificate from Custom CA is invalid: %v", err)
+			return nil, fmt.Errorf("response certificate from Custom CA is invalid: %v", err)
+		}
+		responseCertChains = append(responseCertChains, parsedCert)
+	}
+
+	// Append current roots: Custom CA's root CA and Self signing's root CA
+	rootCertBytes := c.keyCertBundle.GetRootCertPem()
+	responseCertChains = append(responseCertChains, string(rootCertBytes))
+
+	return &pb.IstioCertificateResponse{
+		CertChain: responseCertChains,
+	}, nil
+}
+
+func validateAndParseCert(cert string) (string, error) {
+	defer responseTimeTrack(time.Now(), "validateAndParseCert")
+	certBytes, _ := pem.Decode([]byte(cert))
+
+	if certBytes == nil {
+		return "", fmt.Errorf("decode cert is failed, invalid certificate: %v", cert)
+	}
+
+	block := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes.Bytes,
+	}
+
+	c := pem.EncodeToMemory(block)
+	return string(c), nil
+}
+
+func verifyCSRIdentities(identities []string, csrPEM string) error {
+	if len(identities) == 0 {
+		return nil
+	}
+
+	csrParsed, _ := pem.Decode([]byte(csrPEM))
+	if csrParsed == nil {
+		return fmt.Errorf("decode csr failed - empty pem: %v", csrPEM)
+	}
+	parsedCSR, err := x509.ParseCertificateRequest(csrParsed.Bytes)
+	if err != nil {
+		return fmt.Errorf("cannot parse CSR: %v", err)
+	}
+
+	if len(parsedCSR.Subject.CommonName) != 0 {
+		return fmt.Errorf("received malicious CSR, contains Subject.CommonName: %#v", parsedCSR.Subject.CommonName)
+	}
+
+	if len(parsedCSR.DNSNames) != 0 {
+		return fmt.Errorf("received malicious CSR, contains DNSNames: %v", parsedCSR.DNSNames)
+	}
+
+	if len(parsedCSR.IPAddresses) != 0 {
+		return fmt.Errorf("received malicious CSR, contains IPAddresses: %v", parsedCSR.IPAddresses)
+	}
+
+	idsFromCSR, err := util.ExtractIDs(parsedCSR.Extensions)
```

Comment:
Can I send a CSR with request for `[my-real-identity, root-admin-identity]` and get it signed? seems like we are not doing full checking here.

I am concerned there are other holes in this function. what can we do to ensure this is lockeddown?

Reply:
I checked length of CSR's identities